runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapperTest.java

```diff
@@ -626,13 +650,17 @@ public void testSourceWithNoReaderDoesNotShutdown() throws Exception {
 
       try {
         thread.start();
-        List<UnboundedSource.UnboundedReader<KV<Integer, Integer>>> localReaders =
-            sourceWrapper.getLocalReaders();
-        while (localReaders != null && !localReaders.isEmpty()) {
-          Thread.sleep(200);
-          // should stay alive
-          assertThat(thread.isAlive(), is(true));
+        Thread.sleep(200);
```

Comment:
How did you arrive at this sleep interval?

Reply:
It it just a test that the source does not shut down immediately. The only way to verify is to wait for some amount of time. More importantly the watermark progressing which follows should trigger shutdown immediately if enabled which is verified with `assertThat(sourceWrapper.isRunning(), is(true));`

Reply:
This looks like a recipe for non deterministic test results.. But I don't have any better idea either. Would suggest to add above as comment to the code at least.

Reply:
Comment added.